File: backend/app/services/cache.py
"""
Server-side in-memory cache with TTL support.
"""

from cachetools import TTLCache

# Global cache stores — each with different TTLs
_fixtures_cache = TTLCache(maxsize=500, ttl=3600)       # 1 hour
_standings_cache = TTLCache(maxsize=100, ttl=21600)      # 6 hours
_h2h_cache = TTLCache(maxsize=200, ttl=7200)             # 2 hours
_news_cache = TTLCache(maxsize=50, ttl=3600)             # 1 hour
_live_cache = TTLCache(maxsize=100, ttl=60)              # 1 minute

CACHE_STORES = {
    "fixtures": _fixtures_cache,
    "standings": _standings_cache,
    "h2h": _h2h_cache,
    "news": _news_cache,
    "live": _live_cache,
}


def get_cached(store_name: str, key: str):
    store = CACHE_STORES.get(store_name)
    if store is None:
        return None
    return store.get(key)


def set_cached(store_name: str, key: str, data):
    store = CACHE_STORES.get(store_name)
    if store is not None:
        store[key] = data


def clear_cache(store_name: str = None):
    if store_name:
        store = CACHE_STORES.get(store_name)
        if store:
            store.clear()
    else:
        for store in CACHE_STORES.values():
            store.clear()

File: src/services/cache.ts
/**
 * Simple caching layer using localStorage + in-memory fallback.
 * Supports TTL-based expiration.
 */

const CACHE_PREFIX = 'chamayetupamoja_cache_';

interface CacheEntry<T> {
  data: T;
  timestamp: number;
  ttl: number;
}

// In-memory cache for faster reads within same session
const memoryCache = new Map<string, CacheEntry<any>>();

export function getCached<T>(key: string): T | null {
  const fullKey = CACHE_PREFIX + key;
  
  // Check memory first
  const memEntry = memoryCache.get(fullKey);
  if (memEntry && Date.now() - memEntry.timestamp < memEntry.ttl) {
    return memEntry.data as T;
  }
  
  // Check localStorage
  try {
    const raw = localStorage.getItem(fullKey);
    if (!raw) return null;
    
    const entry: CacheEntry<T> = JSON.parse(raw);
    if (Date.now() - entry.timestamp < entry.ttl) {
      // Restore to memory cache
      memoryCache.set(fullKey, entry);
      return entry.data;
    }
    
    // Expired — clean up
    localStorage.removeItem(fullKey);
    memoryCache.delete(fullKey);
  } catch {
    // Corrupted cache entry
  }
  
  return null;
}

export function setCache<T>(key: string, data: T, ttlMs: number): void {
  const fullKey = CACHE_PREFIX + key;
  const entry: CacheEntry<T> = {
    data,
    timestamp: Date.now(),
    ttl: ttlMs,
  };
  
  memoryCache.set(fullKey, entry);
  
  try {
    localStorage.setItem(fullKey, JSON.stringify(entry));
  } catch {
    // localStorage might be full — memory cache still works
  }
}

export function clearCache(pattern?: string): void {
  if (pattern) {
    const fullPattern = CACHE_PREFIX + pattern;
    // Clear matching keys
    for (const key of Object.keys(localStorage)) {
      if (key.startsWith(fullPattern)) {
        localStorage.removeItem(key);
      }
    }
    for (const key of memoryCache.keys()) {
      if (key.startsWith(fullPattern)) {
        memoryCache.delete(key);
      }
    }
  } else {
    // Clear all cache
    for (const key of Object.keys(localStorage)) {
      if (key.startsWith(CACHE_PREFIX)) {
        localStorage.removeItem(key);
      }
    }
    memoryCache.clear();
  }
}

// Cache TTL constants
export const CACHE_TTL = {
  FIXTURES: 60 * 60 * 1000,        // 1 hour
  STANDINGS: 6 * 60 * 60 * 1000,   // 6 hours  
  LIVE_SCORES: 60 * 1000,          // 1 minute
  NEWS: 60 * 60 * 1000,            // 1 hour
  H2H: 2 * 60 * 60 * 1000,        // 2 hours
} as const;
